### ISSUE TITLE
RAM peaks makes calculation of transition matrix with stochastic mode impossible 

### ISSUE THREAD

**Issue by EmilyMiddendorp** [theislab/cellrank#757]
Hi! 

I am trying to analyse a dataset of about 88k cells. On the Velocity kernel when I try to calculate the transition matrix on stochastic mode, the first step goes well, but the second one gets RAM peak that makes the kernel die every time. I found a similar [question ](https://github.com/theislab/cellrank/issues/466#issuecomment-930021950) on the GitHub but the answer did not solve my problem. Any idea how I can solve this?

**Comment by michalk8** [comment]
Hi @EmilyMiddendorp ,
this is likely due to `jax` and the hessian computation (as far as I remember, it will be of shape `(n_neighbors, n_velo_genes, n_velo_genes)`). A few questions:
- do you use `jax` on GPU? can be checked via 
- how much memory do you have allocated? afaik we did benchmarks (not performance one) on 48k subset of the Morris reprogramming data, but I can't remember how much memory `stochastic` consumed; will try it later on the full 85k data.
- do you see a spike happening (i.e. does it crash immediately after estimating the softmax scale) or does it just gradually increase?
```python
import jax
jax.default_backend()
```
- how many velocity genes are you using? by default, `{vkey}_genes` are used to determine the subset, but you can supply them as `gene_subset` during kernel initialization
- similarly to above, what's the max. number of neighbors of the connectivities? you can check this as 
```python
vk = VelocityKernel(...)
print((vk._conn != 0).sum(1).max())
```
- can you please post the output of ``cr.logging.print_version_and_date()``?

**Comment by EmilyMiddendorp** [comment]
Hi Michal! 

Thank you for the answer. 
So apparently I am running jax on cpu. I got this output to the default_backend: 
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
'cpu'
Can I change that?

I am using a VM with 48G of RAM memory, by the end of step 1 the memory is filled for about 40%. When step 2 starts I see it rapidly increase until it fills up everything (even the Swp memory) and then it dies. This happens even before the widget for the progress bar appears.

I am using 2100 velocity genes, this are genes that I selected to obtain the (biologically) correct embedding in ScVelo. 

The neighbours for the connectivities were set to 500 in the parameters when I calculated it in ScVelo. 

 And this is my cell rank version: Running CellRank 1.4.0, on 2021-09-30 14:21

Do you have any further suggestions on what I can change to be able to run stochastic? Thank you so much for the help!

**Comment by michalk8** [comment]
Thanks for the info!

> So apparently I am running jax on cpu. I got this output to the default_backend:
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
'cpu'
Can I change that?

I wouldn't in this case, since GPUs have much less memory.

> The neighbours for the connectivities were set to 500 in the parameters when I calculated it in ScVelo.

Even if specifying the neighbors to be `500`, it will be more, since the connectivities matrix is symmetrized. E.g. when testing this on the server using `104679` from the Morris reprogramming data, I get:
```python
sc.pp.neighbors(adata, n_pcs=30, n_neighbors=30)
print((adata.obsp['connectivities'] != 0).sum(1).max())  # 323
sc.pp.neighbors(adata, n_pcs=30, n_neighbors=500)
print((adata.obsp['connectivities'] != 0).sum(1).max())  # 3236
```
Because we process the cells in stochastic mode from cells with largest #neighbors to smallest, the first Hessian matrix would in this case need
`3236 (neighs) * 2100 (genes) * 2100 (genes) * 4 (bytes, jax by default is 32bit) / 1024 / 1024 / 1024 = 53GiB` and this does not include the memory need to do the computation, etc. We compute the full Hessian (and only need the diagonal), but as far as I know, it's equally hard to compute the full Hessian or its diagonal (though some approximations exist).

I'd consider these 2 things:
- if you really want/need the `stochastic` mode, either increase the memory on your VM or decrease the #neighbors, if any of these are possible
- if not, you can approximate it with `monte_carlo` estimate (`mode='monte_carlo', n_samples=...`); however, I have no intuition on how many samples you might need

I just tried it on the server with ~`104k` cells, `~19.9GIB` was shown to be used before `jax`, afterwards it jumped to `30.9GiB`, my largest Hessian was to take up `323 * 1135 * 1135 * 4 / 1024 / 1024 / 1024 = ~1.55GiB` of space, but based on the above, seems like this was way too optimistic of an estimate.

**Comment by michalk8** [comment]
Glad to be of help - if there's are any more problems, just let us know. Closing this.
